### ISSUE TITLE
[r/ci] Analyze `arrow` build-from-source on MacOS in R CI

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#2254]
## Problem

Slow CI time (30 extra minutes per run) on MacOS due to build-from-source of the R `arrow` package.

As raised by @eddelbuettel.

## Details

CI runs https://github.com/single-cell-data/TileDB-SOMA/actions/workflows/r-ci.yml

Reading log-files there reveal this leads to "Dependencies" https://github.com/single-cell-data/TileDB-SOMA/blob/1.8.0/.github/workflows/r-ci.yml#L77-L78. This calls https://github.com/single-cell-data/TileDB-SOMA/blob/1.8.0/apis/r/tools/r-ci.sh#L356-L358

Logs:

* Full-output example https://gist.github.com/johnkerl/45747955555e1cb8ac1e5670b9e8f819

* Issue 
  * https://gist.github.com/johnkerl/45747955555e1cb8ac1e5670b9e8f819#file-gistfile1-txt-L3391
  * https://gist.github.com/johnkerl/45747955555e1cb8ac1e5670b9e8f819#file-gistfile1-txt-L3149-L3391

```
2024-03-11T21:24:54.7197580Z   There are binary versions available but the source versions are later:
2024-03-11T21:24:54.7219220Z              binary   source needs_compilation
2024-03-11T21:24:54.7219850Z digest       0.6.34   0.6.35              TRUE
2024-03-11T21:24:54.7220420Z tidyselect    1.2.0    1.2.1              TRUE
2024-03-11T21:24:54.7221570Z arrow      14.0.0.2 14.0.2.1              TRUE
...
2024-03-11T21:26:09.0335010Z *** Building libarrow from source
```

Artifacts:

 * https://cran.r-project.org/package=arrow
 * Shows clearly: macOS binaries:	r-release (arm64): arrow_14.0.0.2.tgz, r-oldrel (arm64): arrow_14.0.0.2.tgz, r-release (x86_64): arrow_14.0.0.2.tgz
* Meanwhile source 14.0.2.1 exists
* In summary: the source version is newer than the latest version for which MacOS binaries are provided

## Solution

TBD. Maybe we need to work with the Arrow folks to provide current MacOS binaries? Or somehow install a pinned version for which binaries _are_ available?

**Comment by eddelbuettel** [comment]
I have not looked and do not have time to look right now but it is almost surely a misconfiguration on our end as
- CRAN always has binaries for both macOS platforms, see https://cloud.r-project.org/package=arrow (and it happens to be skewed today as they admitted 15.0.0.1 to CRAN so builders are catching up
- r-universe has binaries see https://apache.r-universe.dev/arrow (also still at 14.0.2.1)
so I am fairly sure it is just us shooting ourselves into our own foot.  And it didn't used to be the the case -- the three run in r-ci.yaml all came in at 8-ish minutes each.

**Comment by eddelbuettel** [comment]
I also added the key fact that is solely a macOS issue to the Subject.

**Comment by johnkerl** [comment]
FYI as of today -- `libarrow` 15 is out:
https://github.com/single-cell-data/TileDB-SOMA/actions/runs/8257285053/job/22609271820
says:
```
  There is a binary version available but the source version is later:
        binary source needs_compilation
arrow 14.0.0.2 15.0.1              TRUE
```
i.e. MacOS binaries at https://cloud.r-project.org/package=arrow  are now even farther behind source than they were before.

**Comment by eddelbuettel** [comment]
I am aware (I run CRANberries, see below, and read its skeets/toot/RSS) and it is a little orthogonal to my point.  I lack the macOS experience but one just point at one repo and get its binary.  If you do nothing but `install.packages("arrow")` on a mac you should get the most current available build -- be it 14.* or 15.*.  That we end up spending extra time is our doing.

![image](https://github.com/single-cell-data/TileDB-SOMA/assets/673121/b60bf46e-c9f2-4e18-aa9f-6755f606f401)

**Comment by mojaveazure** [comment]
On CRAN, I see the latest macOS binaries for {arrow} are for 14.0.0.2 https://cran.r-project.org/package=arrow while the archive shows that there has also been 14.0.2.1 https://cran.r-project.org/src/contrib/Archive/arrow/?C=M;O=D; this tells me that the macOS binaries have been out-of-date since February 23, 2024

The long compilation time comes from R-arrow source trying to install C++ libarrow from source. On Linux, if a system-wide install of C++ libarrow is already installed, R-arrow source will use that rather than use its own C++ libarrow; I suspect the same is true for macOS (eg. install system-wide libarrow from somewhere, and R-arrow source will install much faster)

Alternatively, we can probably hard-code somewhere `type = "mac.binary"` to force binaries only (default is/should be `type = "both"`)

**Comment by eddelbuettel** [comment]
You are misreading the notion of Archive at CRAN. If you use `available.packages()` pointing at a repo it will never ever read Archives.  It just means a version 14.0.2.1 once existed, but was withdrawn.   I still think we can only get into this confusion by looking at mulitple other sources besides CRAN (and our Additional_Repositories) and then not being careful about not allowing higher source numbers to influence things.  (We had the exact same issue with `bspm` use at times, but for `bspm` I learned the hard way how to toggle between 'binary only, ever' and 'highest version wins, even if source build'.   We have the latter here, and it may be as simple as setting `type` or an equivalent option to `remotes`.

There are also base R alternatives to parsing the DESCRIPTION file info and I am looking into that now.

**Comment by mojaveazure** [comment]
I get that `available.packages()` isn't going to look through the archive. What I'm saying is
- CRAN only has macOS binaries for R-arrow 14.0.0.2
- there has previously been a 14.0.2.1 release of R-arrow (see the Windows `r-release` binary, it's still on 14.0.2.1)
- `bspm` has no role here as this is macOS, not Linux
- R on macOS defaults to `type = "both"`, which will install binaries **_unless_** the source version is newer (as is the case here)

As I said above

> we can probably hard-code somewhere `type = "mac.binary"` to force binaries only (default is/should be `type = "both"`)

**Comment by eddelbuettel** [comment]
There is one authorative source.  It is the CRAN page for the package.  It has binaries.  If we get them from there we have no source build issues.

**Comment by mojaveazure** [comment]
R on macOS defaults to "use binary unless source is newer"; on CRAN, source for R-arrow is newer than binary for macOS and has been since mid/late February. If you grab a Mac and run `install.packages("arrow")`, **_by default_** you'll get arrow 15.0.1 from source because it's newer than the binary (`15.0.1 > 14.0.0.2`)

`remotes` under-the-hood uses the same dependency resolution as `install.pacages()` (it actually calls `utils::install.packages()`), so that is not the problem. The reason we're getting R-arrow from source on macOS, and have been getting R-arrow from source on macOS for ~3 weeks now, is because CRAN does not have, and has not had, up-to-date binaries of R-arrow for macOS. The source version is, and has been, newer than the binary version, so R prefers that

As I said above

> we can probably hard-code somewhere `type = "mac.binary"` to force binaries only (default is/should be `type = "both"`)

Where that gets hard-coded, I haven't looked. But specifying to use binaries only is the answer and works the exact same for both `utils::install.packages()` and `remotes::install_deps()`

**Comment by eddelbuettel** [comment]
Looks like `binary` is accepted as a shorthand too.  From `?options`:

     ‘pkgType’: The default type of packages to be downloaded and
          installed - see install.packages.  Possible values are
          platform dependently

          on Windows ‘"win.binary"’, ‘"source"’ and ‘"both"’ (the
              default).

          on Unix-alikes ‘"source"’ (the default except under a CRAN
              macOS build), ‘"mac.binary"’ and ‘"both"’ (the default
              for CRAN macOS builds).  (‘"mac.binary.el-capitan"’,
              ‘"mac.binary.mavericks"’, ‘"mac.binary.leopard"’ and
              ‘"mac.binary.universal"’ are no longer in use.)

          Value ‘"binary"’ is a synonym for the native binary type (if
          there is one); ‘"both"’ is used by install.packages to
          choose between source and binary installs.

**Comment by eddelbuettel** [comment]
Yep. And the bloody devil is as always hiding in the bloody detail.  We are having some more chatter in parallel but "not there yet".

**Comment by johnkerl** [comment]
Thanks @mojaveazure and @eddelbuettel ! As noted in Slack this is non-urgent, but, I appreciate your expertise and debug here. And everyone benefits from faster CI -- especially during queueing events. 🙏
